File: backend/src/utils/logger.ts
type LogMeta = Record<string, unknown>;

function log(level: "info" | "warn" | "error", message: string, meta?: LogMeta): void {
  const entry = {
    timestamp: new Date().toISOString(),
    level,
    message,
    ...(meta ? { meta } : {})
  };
  const line = JSON.stringify(entry);
  if (level === "error") {
    console.error(line);
  } else if (level === "warn") {
    console.warn(line);
  } else {
    console.log(line);
  }
}

export const logger = {
  info: (message: string, meta?: LogMeta) => log("info", message, meta),
  warn: (message: string, meta?: LogMeta) => log("warn", message, meta),
  error: (message: string, meta?: LogMeta) => log("error", message, meta)
};

File: ticket_router/logger.py
import json
import sys
from datetime import UTC, datetime
from typing import Any


def _emit(level: str, message: str, meta: dict[str, Any] | None = None) -> None:
    entry: dict[str, Any] = {
        "timestamp": datetime.now(UTC).isoformat(),
        "level": level,
        "message": message,
    }
    if meta:
        entry["meta"] = meta
    stream = sys.stderr if level == "error" else sys.stdout
    print(json.dumps(entry, default=str), file=stream)


class Logger:
    """Thin structured-JSON wrapper around plain stdout/stderr logging."""

    def info(self, message: str, meta: dict[str, Any] | None = None) -> None:
        _emit("info", message, meta)

    def warning(self, message: str, meta: dict[str, Any] | None = None) -> None:
        _emit("warning", message, meta)

    def error(self, message: str, meta: dict[str, Any] | None = None) -> None:
        _emit("error", message, meta)


logger = Logger()
